### ISSUE TITLE
Replicating Baidu Transformer Method

### ISSUE THREAD

**Issue by Paulkie99** [SilvioGiancola/SoccerNetv2-DevKit#53]
Hello, I apologise if this isn't the correct place to raise this issue. But seeing as there isn't a public repository available for the [Baidu paper](https://arxiv.org/pdf/2106.14447.pdf) transformer method, I figured there would be more eyes (and willing helpers) here.

I have been trying to replicate the results of the transformer method published above using the [baidu soccer embeddings](https://github.com/baidu-research/vidpress-sports). Unfortunately, it seems that the network is unable to learn meaningful class distinctions and I suspect it is just always predicting the background class. In my experiments using a single dense layer only with the baidu embeddings performs better than when including encoders.

The model summary is given below, I use BCE loss.

![Screenshot 2022-11-13 110225](https://user-images.githubusercontent.com/61508751/201514029-11dec238-d41c-4fb1-9ac2-d7b2c6f4d7c0.png)

Below is the loss curve that I obtained with the LR as stated in the paper (5e-4), batch size 256 -- although I have tried several batch sizes. While the (training) loss does decrease and precision and recall increase briefly, average (training) AUC remains ~0.5. Validation loss also decreases, but all other validation metrics remain constant. When testing, a-mAP is essentially 0. 

While I have mainly used keras, I also modified the NetVLAD++ source code to make use of the transformer model (using [this encoder](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html) and the positional encoder implemented [here](https://pytorch.org/tutorials/beginner/transformer_tutorial.html)). Therefore, I could be sure that my results were not just a data loading issue. I used the following parameters for main.py after implementing the transformer method as a pooling method in model.py called "Baidu". Using these same parameters with pool=NetVLAD++ yields proper results.

`--features=baidu_soccer_embeddings.npy --max_epochs=50 --framerate=1 --window_size=7 --NMS_window=20 --LR=5e-4 --patience=10  --model_name=Baidu --pool=Baidu`

In particular, I added these lines to model.py:

In __init__:

    if not self.input_size == 512:
        if self.input_size == 8576:
            self.feature_extractor = PositionalEncoding(8576)

and at the end of the if/elif chain:

    elif self.pool == "Baidu":
        self.enc = torch.nn.TransformerEncoderLayer(8576, 4, 64)
        self.enc = torch.nn.TransformerEncoder(self.enc, num_layers=3)
        self.fc = nn.Linear(8576, self.num_classes + 1)

In the 'forward' function:

    if not IC == 512:
        if IC == 8576:
            inputs = inputs.reshape(FR, BS, IC)
            inputs = self.feature_extractor(inputs)
            inputs = inputs.reshape(BS, FR, -1)

and at the end of the if/elif chain:

    elif self.pool == "Baidu":
        inputs = self.enc(inputs)
        maxed = torch.max(inputs, 1).values  # also tried flattening
        inputs_pooled = maxed

But the results are similar to those that I obtained with my own code, so I am not sure what the problem could be. While I am not applying mix-up augmentation as stated in the paper, and the training data is less than what they used, I expected the performance to be degraded as a result, but not that I would essentially get 0 a-mAP.

Has anyone else tried to replicate the transformer method? I would appreciate any advice/insights in the name of reproducible research. The intention is to use the transformer method as a baseline in further research.

![loss](https://user-images.githubusercontent.com/61508751/201526727-e2d03471-e67e-4699-8c6a-62a33d6ad90f.png)
![auc](https://user-images.githubusercontent.com/61508751/201526725-abb84723-c3dc-4379-adf0-307bac7a265a.png)

Accuracy and recall plots show similar behaviour to precision below.

![precision](https://user-images.githubusercontent.com/61508751/201526728-3b57ec1a-76ad-4f4b-a99f-ec4701621627.png)

**Comment by Paulkie99** [comment]
I revisited this recently but forgot to update here: the problem was using reshape instead of transpose/permute before applying positional encoding. The shape should also not be altered after positional encoding.
